### PR TITLE
update to MYSQL compatibility

### PR DESCRIPTION
**Pull request by JacobBarthelmeh**
Adds the functions 

>X509_NAME_get_index_by_NID
>X509_NAME_ENTRY_get_data
>ASN1_STRING_data
>ASN1_STRING_length

Implements the previous function stubs

>wolfSSL_X509_NAME_get_entry
>wolfSSL_DES_set_key_unchecked

There is one more function that needs addressed

>wolfSSL_ASN1_TIME_to_string

While tying this function into asn.c code and completing it I want to let the tests run on current code.

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -81,6 +81,7 @@
     #include <wolfssl/wolfcrypt/idea.h>
     #include <wolfssl/wolfcrypt/curve25519.h>
     #include <wolfssl/wolfcrypt/ed25519.h>
+    #include <wolfssl/openssl/asn1.h>
```
Comment (toddouska) [inline src/ssl.c]: openssl/asn1.h should probably go in above block with other openssl compatibility headers.

File: src/ssl.c [hunk 2]
```diff
@@ -3464,13 +3465,35 @@ static int ProcessBuffer(WOLFSSL_CTX* ctx, const unsigned char* buff,
              /* Make sure previous is free'd */
             if (ssl->buffers.weOwnCert) {
                 FreeDer(&ssl->buffers.certificate);
+                #ifdef OPENSSL_EXTRA
+                    FreeX509(ssl->ourCert);
+                    if (ssl->ourCert) {
+                        XFREE(ssl->ourCert, ssl->heap, DYNAMIC_TYPE_X509);
```
Comment (toddouska) [inline src/ssl.c]: after the free of ssl->ourCert this should be set to NULL to prevent double frees.

File: src/ssl.c [hunk 3]
```diff
@@ -3464,13 +3465,35 @@ static int ProcessBuffer(WOLFSSL_CTX* ctx, const unsigned char* buff,
...
+                    }
+                #endif
             }
             XMEMCPY(&ssl->buffers.certificate, &der, sizeof(der));
+            #ifdef OPENSSL_EXTRA
+                ssl->ourCert = wolfSSL_X509_d2i(NULL,
```
Comment (toddouska) [inline src/ssl.c]: Does this processing have to be done now?  That is, could we wait to allocate and process ourCert until it's actually needed?  What about the 99% of people who won't use this?  Isn't that a waste of resources?

**Comment by toddouska** [comment]
wolfSSL_get_certificate() doesn't appear like it will work to me if the cert is loaded into the WOLFSSL object.  Was this tested after the handshake from both a WOLFSSL_CTX cert load and a WOLFSSL cert load?

**Comment by JacobBarthelmeh** [comment]
Thanks for reviewing the code Todd. Loading the cert into the WOLFSSL object was not tested , missed that the weOwnCert variable (and the cert itself) does not stay set for the lifetime of ssl only for during the handshake, will update the pull request to fix this.

File: src/ssl.c [hunk 4]
```diff
@@ -3473,14 +3472,13 @@ static int ProcessBuffer(WOLFSSL_CTX* ctx, const unsigned char* buff,
...
+                        ssl->ourCert = NULL;
                     }
                 #endif
             }
             XMEMCPY(&ssl->buffers.certificate, &der, sizeof(der));
             #ifdef OPENSSL_EXTRA
-                ssl->ourCert = wolfSSL_X509_d2i(NULL,
-                                              ssl->buffers.certificate->buffer,
-                                              ssl->buffers.certificate->length);
+                ssl->keepCert = 1; /* hold cert for ssl lifetime */
```
Comment (toddouska) [inline src/ssl.c]: So every user now has to "pay" for this on every connection?  Keeping memory and resources for something that 99% probably will never use.  I'm having a lot of trouble justifying that.

**Comment by toddouska** [comment]
Thanks for the changes, looks good.

**Pull request merged**